server/src/test/java/org/opensearch/action/search/AbstractSearchAsyncActionTests.java

```diff
@@ -528,9 +548,339 @@ public void onFailure(Exception e) {
         assertThat(searchResponse.getSuccessfulShards(), equalTo(shards.length));
     }
 
+    public void testSearchRequestListeners() {
+        SearchCoordinatorStatsTesting searchCoordinatorStatsTesting = new SearchCoordinatorStatsTesting();
+        SearchRequestOperationsListener testListener = createSearchRequestOperationsListener(searchCoordinatorStatsTesting);
+        final List<SearchRequestOperationsListener> requestOperationListeners = new ArrayList<>(Arrays.asList(testListener, testListener));
+
+        SearchQueryThenFetchAsyncAction action = createSearchQueryThenFetchAsyncAction();
+        action.setSearchListenerList(requestOperationListeners);
+
+        SearchDfsQueryThenFetchAsyncAction searchDfsQueryThenFetchAsyncAction = createSearchDfsQueryThenFetchAsyncAction();
+        searchDfsQueryThenFetchAsyncAction.setSearchListenerList(requestOperationListeners);
+
+        CanMatchPreFilterSearchPhase canMatchPreFilterSearchPhaseAction = createCanMatchPreFilterSearchPhase();
+        canMatchPreFilterSearchPhaseAction.setSearchListenerList(requestOperationListeners);
+
+        action.start();
+        assertEquals(2, searchCoordinatorStatsTesting.queryPhaseStart.get());
+
+        FetchSearchPhase fetchPhase = createFetchSearchPhase();
+        ShardId shardId = new ShardId(randomAlphaOfLengthBetween(5, 10), randomAlphaOfLength(10), randomInt());
+        SearchShardIterator searchShardIterator = new SearchShardIterator(null, shardId, Collections.emptyList(), OriginalIndices.NONE);
+        searchShardIterator.resetAndSkip();
+        action.skipShard(searchShardIterator);
+        action.executeNextPhase(action, fetchPhase);
+
+        assertEquals(2, searchCoordinatorStatsTesting.queryPhaseEnd.get());
+
+        searchDfsQueryThenFetchAsyncAction.start();
+        searchDfsQueryThenFetchAsyncAction.skipShard(searchShardIterator);
+        searchDfsQueryThenFetchAsyncAction.executeNextPhase(searchDfsQueryThenFetchAsyncAction, fetchPhase);
+
+        canMatchPreFilterSearchPhaseAction.start();
+
+        assertEquals(2, searchCoordinatorStatsTesting.dfsPreQueryPhaseStart.get());
+        assertEquals(2, searchCoordinatorStatsTesting.dfsPreQueryPhaseEnd.get());
+        assertEquals(2, searchCoordinatorStatsTesting.canMatchPhaseStart.get());
+        assertEquals(4, searchCoordinatorStatsTesting.fetchPhaseStart.get());
+
+        ExpandSearchPhase expandPhase = createExpandSearchPhase();
+        action.executeNextPhase(fetchPhase, expandPhase);
+        assertEquals(2, searchCoordinatorStatsTesting.expandPhaseStart.get());
+
+        action.sendSearchResponse(InternalSearchResponse.empty(), null);
+    }
+
+    private SearchDfsQueryThenFetchAsyncAction createSearchDfsQueryThenFetchAsyncAction() {
```

Comment:
See if you can utilize individual SearchDfsQueryThenFetchAsyncActionTests for these kind of scenarios. Applies same for rest.